internal/worker/kv_status.go

```diff
@@ -89,3 +91,94 @@ func NewStatusKVPublisher(s events.Stream, log *logrus.Logger, opts ...kv.Option
 		log: log,
 	}
 }
+
+type taskState int
+
+const (
+	notStarted    taskState = iota
+	inProgress              // another flasher has started it, is still around and updated recently
+	complete                // task is done
+	orphaned                // the flasher that started this task doesn't exist anymore
+	indeterminate           // we got an error in the process of making the check
+)
+
+func (o *Worker) taskInProgress(cID string) taskState {
+	handle, err := events.AsNatsJetStreamContext(o.stream.(*events.NatsJetstream)).KeyValue(statusKVName)
+	if err != nil {
+		o.logger.WithError(err).WithFields(logrus.Fields{
+			"condition_id": cID,
+		}).Warn("unable to connect to status KV for condition lookup")
+		return indeterminate
+	}
+
+	lookupKey := fmt.Sprintf("%s.%s", o.facilityCode, cID)
+	entry, err := handle.Get(lookupKey)
+	switch err {
+	case nats.ErrKeyNotFound:
+		// This should be by far the most common path through this code.
+		return notStarted
+	case nil:
+		break // we'll handle this outside the switch
+	default:
+		o.logger.WithError(err).WithFields(logrus.Fields{
+			"condition_id": cID,
+		}).Warn("error reading condition status")
+		return indeterminate
+	}
+
+	// we have an status entry for this condition, is is complete?
+	sv := types.StatusValue{}
+	if errJson := json.Unmarshal(entry.Value(), &sv); errJson != nil {
+		o.logger.WithError(errJson).WithFields(logrus.Fields{
+			"condition_id": cID,
+		}).Warn("unable to construct a sane status for this condition")
+		return indeterminate
+	}
+
+	if cotyp.ConditionState(sv.State) == cotyp.Failed ||
+		cotyp.ConditionState(sv.State) == cotyp.Succeeded {
+		o.logger.WithFields(logrus.Fields{
+			"condition_id":    cID,
+			"condition_state": sv.State,
+		}).Info("this condition is already complete")
+		return complete
+	}
+
+	// is the worker handling this condition alive?
+	worker, err := registry.ControllerIDFromString(sv.WorkerID)
+	if err != nil {
+		o.logger.WithError(err).WithFields(logrus.Fields{
+			"condition_id": cID,
+			"worker_id":    sv.WorkerID,
+		}).Warn("bad worker id")
+		return indeterminate
+	}
+
+	activeAt, err := registry.LastContact(worker)
+	switch err {
+	case nats.ErrKeyNotFound:
+		// the data for this worker aged-out, it's no longer active
+		// XXX: the most conservative thing to do here is to return
+		// indeterminate but most times this will indicate that the
+		// worker crashed/restarted and this task should be restarted.
+		o.logger.WithFields(logrus.Fields{
+			"condition_id": cID,
+			"worker_id":    sv.WorkerID,
+		}).Info("original worker not found")
+		return orphaned
+	case nil:
+		timeStr, _ := activeAt.MarshalText()
```

Comment:
could the `timeStr` value be logged in each of the other case statements as well?

Reply:
We can't. In the case where we return a non-`nil` error, the time value is going to be the zero-value of `time.Time`